libbpf-tools/softirqs.c

```diff
@@ -0,0 +1,277 @@
+"    softirqss -NT 1      # 1s summaries, nanoseconds, and timestamps\n";
+
+static const struct argp_option opts[] = {
+	{ "distributed", 'd', NULL, 0, "Show distributions as histograms" },
+	{ "timestamp", 'T', NULL, 0, "Include timestamp on output" },
+	{ "nanoseconds", 'N', NULL, 0, "Output in nanoseconds" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	static int pos_args;
+
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'd':
+		env.distributed = true;
+		break;
+	case 'N':
+		env.nanoseconds = true;
+		break;
+	case 'T':
+		env.timestamp = true;
+		break;
+	case ARGP_KEY_ARG:
+		errno = 0;
+		if (pos_args == 0) {
+			env.interval = strtol(arg, NULL, 10);
+			if (errno) {
+				fprintf(stderr, "invalid internal\n");
+				argp_usage(state);
+			}
+		} else if (pos_args == 1) {
+			env.times = strtol(arg, NULL, 10);
+			if (errno) {
+				fprintf(stderr, "invalid times\n");
+				argp_usage(state);
+			}
+		} else {
+			fprintf(stderr,
+				"unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		pos_args++;
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+	exiting = true;
+}
+
+enum {
+	HI_SOFTIRQ = 0,
+	TIMER_SOFTIRQ = 1,
+	NET_TX_SOFTIRQ = 2,
+	NET_RX_SOFTIRQ = 3,
+	BLOCK_SOFTIRQ = 4,
+	IRQ_POLL_SOFTIRQ = 5,
+	TASKLET_SOFTIRQ = 6,
+	SCHED_SOFTIRQ = 7,
+	HRTIMER_SOFTIRQ = 8,
+	RCU_SOFTIRQ = 9,
+	NR_SOFTIRQS = 10,
+};
+
+static char *vec_names[] = {
+	[HI_SOFTIRQ] = "hi",
+	[TIMER_SOFTIRQ] = "timer",
+	[NET_TX_SOFTIRQ] = "net_tx",
+	[NET_RX_SOFTIRQ] = "net_rx",
+	[BLOCK_SOFTIRQ] = "block",
+	[IRQ_POLL_SOFTIRQ] = "irq_poll",
+	[TASKLET_SOFTIRQ] = "tasklet",
+	[SCHED_SOFTIRQ] = "sched",
+	[HRTIMER_SOFTIRQ] = "hrtimer",
+	[RCU_SOFTIRQ] = "rcu",
+};
+
+static int print_count(int fd)
+{
+	char *units = env.nanoseconds ? "nsecs" : "usecs";
+	__u64 count, zero = 0;
+	__u32 vec;
+	int err;
+
+	printf("%-16s %6s%5s\n", "SOFTIRQ", "TOTAL_", units);
+
+	for (vec = 0; vec < NR_SOFTIRQS; vec++) {
+		err = bpf_map_lookup_elem(fd, &vec, &count);
+		if (err < 0) {
+			fprintf(stderr, "failed to lookup count: %d\n", err);
+			return -1;
+		}
+		if (count > 0)
+			printf("%-16s %11llu\n", vec_names[vec], count);
+	}
+
+	for (vec = 0; vec < NR_SOFTIRQS; vec++) {
+		err = bpf_map_update_elem(fd, &vec, &zero, 0);
+		if (err < 0) {
+			fprintf(stderr, "failed to cleanup count : %d\n", err);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+static int print_hist(int fd)
+{
+	char *units = env.nanoseconds ? "nsecs" : "usecs";
+	struct hist hist, zero = {};
+	__u32 vec;
+	int err;
+
+	for (vec = 0; vec < NR_SOFTIRQS; vec++) {
+		err = bpf_map_lookup_elem(fd, &vec, &hist);
```

Comment:
given `NR_SOFTIRQS` is small and fixed, there is no reason to use ARRAY map, just use global array variable for this. It will be faster and simpler on both BPF and user-space side.

Reply:
Fixed, thanks!